### PR TITLE
EZP-23818 : As an editor, I want to be able to fill the RelationList field

### PR DESCRIPTION
**Pull request by StephaneDiot**
## Description

The goal here is to be able to edit and create a content with the relation list field
ScreenCast : http://youtu.be/tUBIx6tKaCc

## Link

Jira : https://jira.ez.no/browse/EZP-23818

### CODE REVIEW

File: Resources/public/css/theme/views/fields/edit/relationlist.css [hunk 1]
```diff
@@ -0,0 +1,26 @@
+/**
+ * Copyright (C) eZ Systems AS. All rights reserved.
+ * For full copyright and license information view LICENSE file distributed with this source code.
+ */
+
+.ez-view-relationlisteditview .ez-relation-properties {
```
Comment (dpobel) [inline Resources/public/css/theme/views/fields/edit/relationlist.css]: copy pasted from Relation edit view CSS and useless here

File: Resources/public/css/views/fields/edit/relationlist.css [hunk 2]
```diff
@@ -0,0 +1,78 @@
+/**
+ * Copyright (C) eZ Systems AS. All rights reserved.
+ * For full copyright and license information view LICENSE file distributed with this source code.
+ */
+
+.ez-view-relationlisteditview .ez-relation-content {
```
Comment (dpobel) [inline Resources/public/css/views/fields/edit/relationlist.css]: as the previous CSS file, too much copy paste, please clean the CSS

File: Resources/public/js/views/fields/ez-relationlist-editview.js [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+            },
+            '.ez-relation-remove-content': {
+                'tap': '_removeRelation'
+            }
+        },
+
+        initializer: function () {
+            this._fireMethod = this._fireLoadFieldRelatedContents;
+            this._watchAttribute = 'destinationContents';
+            if( this.get('field').fieldValue.destinationContentIds ){
```
Comment (dpobel) [inline Resources/public/js/views/fields/ez-relationlist-editview.js]: put `this.get('field').fieldValue` in a variable

File: Resources/public/js/views/fields/ez-relationlist-editview.js [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+     * Provides the field edit view for the relation list fields
+     *
+     * @module ez-relationlist-editview
+     */
+    Y.namespace('eZ');
+
+    var FIELDTYPE_IDENTIFIER = 'ezobjectrelationlist';
+
+    /**
+     * Relation edit view
```
Comment (dpobel) [inline Resources/public/js/views/fields/ez-relationlist-editview.js]: to be updated

File: Resources/public/js/views/fields/ez-relationlist-editview.js [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+                this._set('destinationContentsIds', this.get('field').fieldValue.destinationContentIds);
+            }
+            this.on('destinationContentsChange', this._syncDestinationContentsIds);
+        },
+
+        /**
+         * Synchronize the destinationContentId attribute when destinationContent change.
+         *
+         * @method _syncDestinationContentsIds
+         * @protected
```
Comment (dpobel) [inline Resources/public/js/views/fields/ez-relationlist-editview.js]: missing the parameter doc

File: Resources/public/js/views/fields/ez-relationlist-editview.js [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+
+        /**
+         * Checks whether the field is empty
+         *
+         * @method _isFieldEmpty
+         * @protected
+         * @return {Boolean}
+         */
+        _isFieldEmpty: function () {
+            /*console.log(!this.get('destinationContentsIds'));
```
Comment (dpobel) [inline Resources/public/js/views/fields/ez-relationlist-editview.js]: to be removed

File: Resources/public/js/views/fields/ez-relationlist-editview.js [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+         * Returns an object containing the additional variables
+         *
+         * @method _variables
+         * @protected
+         * @return Object
+         */
+        _variables: function () {
+            var dest = this.get('destinationContents'),
+                destinationContentsJSON = [];
+            if ( dest.length > 0 ) {
```
Comment (dpobel) [inline Resources/public/js/views/fields/ez-relationlist-editview.js]: this condition is useless

File: Resources/public/js/views/fields/ez-relationlist-editview.js [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+            return {
+                destinationContents:  destinationContentsJSON,
+                loadingError: this.get('loadingError'),
+                isEmpty: this._isFieldEmpty(),
+                isRequired: this.get('fieldDefinition').isRequired,
+            };
+        },
+
+        /**
+         * Tap event handler for the remove button. It resets the relation.
```
Comment (dpobel) [inline Resources/public/js/views/fields/ez-relationlist-editview.js]: to be updated

File: Resources/config/yui.yml [hunk 9]
```diff
@@ -468,6 +469,12 @@ system:
                 relationeditview-ez-template:
                     type: 'template'
                     path: %ez_platformui.public_dir%/templates/fields/edit/relation.hbt
+                ez-relationlist-editview:
+                    requires: ['ez-fieldeditview', 'relationlisteditview-ez-template', 'ez-asynchronousview']
```
Comment (dpobel) [inline Resources/config/yui.yml]: it requires `array-extras` as well (for `Y.Array.reject`)

File: Resources/public/js/views/fields/ez-relationlist-editview.js [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+         *
+         * @method _removeRelation
+         * @protected
+         * @param {EventFacade} e
+         */
+        _removeRelation: function (e) {
+            var temp;
```
Comment (dpobel) [inline Resources/public/js/views/fields/ez-relationlist-editview.js]: `temp` is a very bad name, `remainingContents` would be better for instance.

File: Resources/public/js/views/fields/ez-relationlist-editview.js [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+                config: {
+                    multiple: true,
+                    contentDiscoveredHandler: Y.bind(this._selectRelation, this),
+                    cancelDiscoverHandler: Y.bind(this.validate, this),
+                },
+            });
+        },
+
+        /**
+         * Universal discovery contentDiscovered event handler to fill the relation after the user chose a content
```
Comment (dpobel) [inline Resources/public/js/views/fields/ez-relationlist-editview.js]: the user can choose several contents, so this comment needs to be updated

File: Resources/public/js/views/fields/ez-relationlist-editview.js [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+         * @return Object
+         */
+        _getFieldValue: function () {
+            this.validate();
+            return {destinationContentIds: this.get('destinationContentsIds')};
+        },
+    },{
+        ATTRS: {
+            /**
+             * The destination content of the relation
```
Comment (dpobel) [inline Resources/public/js/views/fields/ez-relationlist-editview.js]: to be updated

File: Resources/public/templates/fields/edit/relationlist.hbt [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+            <p class="ez-editfield-error-message">&nbsp;</p>
+        </label>
+    </div>
+    <div class="pure-u ez-editfield-input-area ez-default-error-style">
+        <div class="ez-editfield-input"><div class="ez-relation-input-ui">
+            {{#if isEmpty}}
+                <p class="ez-relation-empty">No relation.</p>
+            {{else}}
+                {{#if destinationContents }}
+                <table border="1">
```
Comment (dpobel) [inline Resources/public/templates/fields/edit/relationlist.hbt]: remove the `border="1"`, this should be handled in the CSS. For a nicer table, you can also use the PureCSS style http://purecss.io/tables/ (it's used in https://github.com/ezsystems/PlatformUIBundle/blob/master/Resources/views/Section/list.html.twig#L54 for instance)

File: Resources/public/templates/fields/edit/relationlist.hbt [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+                <tr><th>Title</th><th>Published</th><th>Modified</th><th>Remove relation</th></tr>
```
Comment (dpobel) [inline Resources/public/templates/fields/edit/relationlist.hbt]: please indent

File: Resources/public/templates/fields/edit/relationlist.hbt [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+                    {{#each destinationContents}}
+                        <tr>
+                            <td class="ez-relation-content-icon" data-icon="&#xe601;"><div class="ez-relation-content-name"><strong>{{ name }}</strong></div></td>
```
Comment (dpobel) [inline Resources/public/templates/fields/edit/relationlist.hbt]: this is over complicated.
```html
<td class="ez-relation-content-name" data-icon="&#xe601;">{{ name }}</td>
```
is enough to me

File: Resources/public/templates/fields/edit/relationlist.hbt [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+                            <td class="ez-relation-property">{{ publishedDate }}</td>
+                            <td class="ez-relation-property">{{ lastModificationDate }}</td>
+                                <td class="ez-relation-remove-content"><button data-content-id="{{id}}" class=" ez-button ez-button-delete ez-font-icon pure-button">trash</button></td>
```
Comment (dpobel) [inline Resources/public/templates/fields/edit/relationlist.hbt]: Please indent correctly.
Why *trash* as the label of the button ? *Remove*

File: Resources/public/templates/fields/edit/relationlist.hbt [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+                    </table>
+
+                {{else}}
+                    {{#if loadingError}}
+                        <p class="ez-asynchronousview-error ez-font-icon">
+                            An error occurred while loading the related content.
+                            <button class="ez-asynchronousview-retry ez-button ez-font-icon pure-button">Retry</button>
+                        </p>
+                    {{else}}
+                        <p class="ez-font-icon ez-asynchronousview-loading">Loading the related content</p>
```
Comment (dpobel) [inline Resources/public/templates/fields/edit/relationlist.hbt]: this needs to be updated

File: Tests/js/views/fields/assets/ez-relationlist-editview-tests.js [hunk 18]
```diff
@@ -0,0 +1,569 @@
...
+            this.view.template = function (variables) {
+                templateCalled = true;
+                return origTpl.apply(this, arguments);
+            };
+            this.view.set('loadingError', true);
+            Y.Assert.isTrue(templateCalled, "The template has not been used");
+        },
+
+
+        "Should fill the destinationContentsIds attribute from the field": function () {
```
Comment (dpobel) [inline Tests/js/views/fields/assets/ez-relationlist-editview-tests.js]: this test should be put in a separate test case

File: Tests/js/views/fields/assets/ez-relationlist-editview-tests.js [hunk 19]
```diff
@@ -0,0 +1,569 @@
...
+            for (var i = 0; i <= this.view.get('destinationContentsIds').length; i++) {
+                Y.Assert.areSame(
+                    this.view.get('field').fieldValue.destinationContentIds[i],
+                    this.view.get('destinationContentsIds')[i],
+                    "The destinationContentId of the field value should be the same than the attribute"
+                );
+            }
+        },
+
+        "Should fire the loadFieldRelatedContents event": function () {
```
Comment (dpobel) [inline Tests/js/views/fields/assets/ez-relationlist-editview-tests.js]: different test case as well

File: Tests/js/views/fields/assets/ez-relationlist-editview-tests.js [hunk 20]
```diff
@@ -0,0 +1,569 @@
...
+                    "fieldDefinitionIdentifier is the same than the one in the field"
+                );
+                loadContentEvent = true;
+            });
+            this.view.set('active', true);
+
+            Y.Assert.isTrue(loadContentEvent, "loadContentEvent should be called when changing active value");
+        },
+
+        "Should NOT fire the loadFieldRelatedContents event if field is empty": function () {
```
Comment (dpobel) [inline Tests/js/views/fields/assets/ez-relationlist-editview-tests.js]: should go in the same test case as "Should fire the loadFieldRelatedContents event"

File: Tests/js/views/fields/assets/ez-relationlist-editview-tests.js [hunk 21]
```diff
@@ -0,0 +1,569 @@
...
+                );
+                loadContentEvent = true;
+            });
+            this.view._set('destinationContentsIds', null);
+            this.view.set('active', true);
+
+            Y.Assert.isFalse(loadContentEvent, "loadContentEvent should NOT be called when changing active value");
+        },
+
+        "Should prevent default behaviour of the tap event for select button": function () {
```
Comment (dpobel) [inline Tests/js/views/fields/assets/ez-relationlist-editview-tests.js]: the following two tests should go in their own test case

File: Tests/js/views/fields/assets/ez-relationlist-editview-tests.js [hunk 22]
```diff
@@ -0,0 +1,569 @@
...
+        name: "eZ Relation list View test",
+
+        _getFieldDefinition: function (required) {
+            return {
+                isRequired: required
+            };
+        },
+
+        setUp: function () {
+
```
Comment (dpobel) [inline Tests/js/views/fields/assets/ez-relationlist-editview-tests.js]: pretty sure this method can be stripped down, for instance is it necessary to mock the `get` methods of the destination contents ?

**Comment by StephaneDiot** [comment]
ping @dpobel

File: Resources/public/js/views/fields/ez-relationlist-editview.js [hunk 23]
```diff
@@ -0,0 +1,230 @@
...
+
+    /**
+     * Relation list edit view
+     *
+     * @namespace eZ
+     * @class RelationListEditView
+     * @constructor
+     * @extends eZ.FieldEditView
+     */
+    Y.eZ.RelationListEditView = Y.Base.create('relationListEditView', Y.eZ.FieldEditView, [Y.eZ.AsynchronousView], {
```
Comment (yannickroger) [inline Resources/public/js/views/fields/ez-relationlist-editview.js]: nitpick: line is too long

File: Resources/public/js/views/fields/ez-relationlist-editview.js [hunk 24]
```diff
@@ -0,0 +1,230 @@
...
+                });
+            }
+        },
+
+        /**
+         * Checks whether the field is empty
+         *
+         * @method _isFieldEmpty
+         * @protected
+         * @return {Boolean}
```
Comment (yannickroger) [inline Resources/public/js/views/fields/ez-relationlist-editview.js]: I don't think you need jsdoc when method is "inherited".

File: Resources/public/js/views/fields/ez-relationlist-editview.js [hunk 25]
```diff
@@ -0,0 +1,230 @@
...
+                config: {
+                    multiple: true,
+                    contentDiscoveredHandler: Y.bind(this._selectRelation, this),
+                    cancelDiscoverHandler: Y.bind(this.validate, this),
+                },
+            });
+        },
+
+        /**
+         * Universal discovery contentDiscovered event handler to fill the relation list after the user chose one or several contents.
```
Comment (yannickroger) [inline Resources/public/js/views/fields/ez-relationlist-editview.js]: nitpick: line too long

File: Tests/js/views/fields/assets/ez-relationlist-editview-tests.js [hunk 26]
```diff
@@ -0,0 +1,763 @@
...
+        name: "eZ Relation list universal discovery relation test",
+
+        _getFieldDefinition: function (required) {
+            return {
+                isRequired: required
+            };
+        },
+
+        setUp: function () {
+
```
Comment (yannickroger) [inline Tests/js/views/fields/assets/ez-relationlist-editview-tests.js]: extra line

**Pull request merged**